library/src/main/api21/com/google/android/cameraview/Camera2.java

```diff
@@ -435,11 +450,49 @@ protected void collectPictureSizes(SizeMap sizes, StreamConfigurationMap map) {
 
     private void prepareImageReader() {
         Size largest = mPictureSizes.sizes(mAspectRatio).last();
-        mImageReader = ImageReader.newInstance(largest.getWidth(), largest.getHeight(),
+        initImageReader(largest.getWidth(), largest.getHeight());
+
+        Size size = chooseOptimalSize();
+        initCallbackImageDataReader(size.getWidth(), size.getHeight());
+    }
+
+    private void initImageReader(int width, int height){
+        mImageReader = ImageReader.newInstance(width, height,
                 ImageFormat.JPEG, /* maxImages */ 2);
         mImageReader.setOnImageAvailableListener(mOnImageAvailableListener, null);
     }
 
+    private void initCallbackImageDataReader(int width, int height) {
+
+        if (mCallbackImageDataReader != null) {
+            mCallbackImageDataReader.close();
+        }
+
+        mCallbackImageDataReader = ImageReader.newInstance(width, height,
+                ImageFormat.NV21, /* maxImages */ 1);
+        mCallbackImageDataReader.setOnImageAvailableListener(mOnPreviewImageAvailableListener, null);
+    }
+
+    /**
+     * Extract the byte data from the ImageReader
+     * @param reader
+     * @return byte array or null if failed
+     */
+    @Nullable
+    private byte[] getByteDataFromImageReader(@NonNull ImageReader reader) {
+        byte[] data = null;
+        try (Image image = reader.acquireNextImage()) {
+            Image.Plane[] planes = image.getPlanes();
+            if (planes.length > 0) {
+                ByteBuffer buffer = planes[0].getBuffer();
+                data = new byte[buffer.remaining()];
+                buffer.get(data);
+            }
+        }
+        return data;
+    }
+
```

Comment:
Some image formats have multiple planes, including YUV420_888 (the default for Camera2).  All planes need to be included in the output byte array.